samples/net/ptp/slave/src/ptp_slave.c

```diff
@@ -0,0 +1,313 @@
+	int64_t t1_t4_avg_ns = (ptp_ts_net_to_ns(&state.t1) + ptp_ts_net_to_ns(&state.t4)) / 2;
+	int64_t t2_t3_avg_ns = (ptp_ts_net_to_ns(&state.t2) + ptp_ts_net_to_ns(&state.t3)) / 2;
+	int64_t offset_ns = t1_t4_avg_ns - t2_t3_avg_ns;
+	int ret;
+
+	LOG_DBG("t1: %lu.%09u, t2: %lu.%09u, t3: %lu.%09u, t4: %lu.%09u, "
+		"delay: %d, offset: %c%lu.%09u",
+		(unsigned long)state.t1.second, state.t1.nanosecond,
+		(unsigned long)state.t2.second, state.t2.nanosecond,
+		(unsigned long)state.t3.second, state.t3.nanosecond,
+		(unsigned long)state.t4.second, state.t4.nanosecond,
+		delay_ns, offset_ns >= 0 ? '+' : '-', labs(offset_ns / NSEC_PER_SEC),
+		abs(offset_ns % NSEC_PER_SEC));
+
+	if (llabs(offset_ns) > NSEC_PER_SEC / 2) {
+		/*
+		 * We'll assume that good estimate of current time is:
+		 * t4 + t2_t3_delay_ns / 2 + delay_ns
+		 */
+		int64_t current_ns = ptp_ts_net_to_ns(&state.t4) + t2_t3_delay_ns / 2 + delay_ns;
+		struct net_ptp_time current_t;
+
+		current_t.second = current_ns / NSEC_PER_SEC;
+		current_t.nanosecond = current_ns % NSEC_PER_SEC;
+
+		/* time delta is too large - jump */
+		ret = ptp_clock_set(state.clk, &current_t);
+		if (ret < 0) {
+			LOG_WRN("ptp_clock_set() failed: %d", ret);
+			return;
+		}
+		LOG_INF("PTP clock set: %lu.%09u", (unsigned long)current_t.second,
+			current_t.nanosecond);
+		state.last_offset_ns = 0;
+		return;
+	}
+
+	/* compensate the offset using a PD controller */
+	double rate = 1.0 + state.kp * (double)offset_ns / NSEC_PER_SEC +
+		      state.kd * (double)(offset_ns - state.last_offset_ns) / NSEC_PER_SEC;
+	ret = ptp_clock_rate_adjust(state.clk, rate);
+	if (ret < 0) {
+		LOG_WRN("ptp_clock_rate_adjust() failed: %d", ret);
+		return;
+	}
+	state.last_offset_ns = offset_ns;
+}
+
+static void pkt_received(struct net_context *context, struct net_pkt *pkt,
+			 union net_ip_header *ip_hdr, union net_proto_header *proto_hdr,
+			 int status, void *user_data)
+{
+	if (ntohs(NET_ETH_HDR(pkt)->type) != NET_ETH_PTYPE_PTP) {
+		goto done;
+	}
+
+	/* skip ethernet header */
+	net_pkt_skip(pkt, sizeof(struct net_eth_hdr));
+
+	/* read PTP header */
+	struct ptp_header ptp_hdr;
+
+	net_pkt_read(pkt, &ptp_hdr, sizeof(ptp_hdr));
+
+	if (ptp_hdr.version_ptp != 2 || ptp_hdr.minor_version_ptp != 0) {
+		LOG_WRN("Unsupposed PTP v%d.%d packet ignored", ptp_hdr.version_ptp,
+			ptp_hdr.minor_version_ptp);
+		goto done;
+	}
+
+	switch (ptp_hdr.message_type) {
+	case PTP_MESSAGE_TYPE_SYNC:
+		handle_ptp_sync(pkt, &ptp_hdr);
+		break;
+	case PTP_MESSAGE_TYPE_FOLLOW_UP:
+		handle_ptp_follow_up(pkt, &ptp_hdr);
+		break;
+	case PTP_MESSAGE_TYPE_DELAY_RESP:
+		handle_ptp_delay_resp(pkt, &ptp_hdr);
+		break;
+	}
+
+done:
+	net_pkt_unref(pkt);
+}
+
+int ptp_ts_get(struct net_ptp_time *net)
+{
+	return ptp_clock_get(state.clk, net);
+}
+
+int ptp_slave_init(struct net_if *iface)
+{
+	int ret;
+
+	memset(&state, 0, sizeof(state));
+
+	state.iface = iface;
+
+	state.clk = net_eth_get_ptp_clock(iface);
```

Comment:
Please check that the returned value is not NULL, as it is possible that the interface does not support PTP (if there are multiple Ethernet interfaces in the system).

Reply:
Addressed in #41031